## KNN - Plane Total Landed Weight

K Nearest Neighbours algorithm using San Francisco air traffic dataset. Trying to predict the total landed weight of a plane.

In [1]:
# Libraries
import pandas as pd
import numpy as np

import sklearn
from sklearn.utils import shuffle
from sklearn.neighbors import KNeighborsClassifier
from sklearn import linear_model, preprocessing


In [2]:
air_t = pd.read_csv('air-traffic-landings-statistics_v2.csv')

In [3]:
# check the dimesions of dataset
air_t.shape

(21762, 10)

In [4]:
# View first rows of dataset
air_t.head()

Operating Airline Operating Airline IATA Code    GEO Summary GEO Region  \
0      ATA Airlines                          TZ       Domestic         US   
1      ATA Airlines                          TZ       Domestic         US   
2      ATA Airlines                          TZ       Domestic         US   
3        Air Canada                          AC  International     Canada   
4        Air Canada                          AC  International     Canada   

  Landing Aircraft Type Aircraft Body Type Aircraft Manufacturer  \
0             Passenger        Narrow Body                Boeing   
1             Passenger        Narrow Body                Boeing   
2             Passenger          Wide Body              Lockheed   
3             Passenger        Narrow Body                Boeing   
4             Passenger        Narrow Body                Boeing   

  Aircraft Model  Landing Count  Total Landed Weight  
0            757             83             16434000  
1            757              3               672000  
2          L1011             27              9666000  
3            737              5               525000  
4            737             15              1605000

The dataset consists of 10 features, 2 of which are numerical, 8 categorical.

Last column 'Total Landed Weight' will be set as the target variable.

In [5]:
# view stats of dataset
air_t.describe()

Landing Count  Total Landed Weight
count   21762.000000         2.176200e+04
mean      113.421652         1.896583e+07
std       248.910829         3.009876e+07
min         1.000000         6.850000e+03
25%        14.000000         3.080500e+06
50%        31.000000         9.678039e+06
75%        84.000000         1.953000e+07
max      2245.000000         2.730420e+08

In [6]:
# Features by type:
# Operating Airline: Categorical
# Operating Airline IATA Code: Categorical
# GEO Summary: Categorical
# GEO Region: Categorical
# Landing Aircraft Type: Categorical
# Aircraft Body Type: Categorical
# Aircraft Manufacturer: Categorical
# Aircraft Model: Categorical
# Landing Count: Numerical
# Total Landed Weight: Numerical

In [8]:
# Need to use non-numerical data for variables:
# sklearn will covert non-numerical values into numerical values.

le = preprocessing.LabelEncoder()    # coding the labels into integer values (using Label Encoder function)

# creating lists for each column:
Operating_Airline = le.fit_transform(list(air_t['Operating Airline']))
Op_IATA_Code = le.fit_transform(list(air_t['Operating Airline IATA Code']))
GEO_Summary = le.fit_transform(list(air_t['GEO Summary']))
GEO_Region = le.fit_transform(list(air_t['GEO Region']))
Landing_Aircraft_Type = le.fit_transform(list(air_t['Landing Aircraft Type']))
Aircraft_Body_Type = le.fit_transform(list(air_t['Aircraft Body Type']))
Aircraft_Manufacturer = le.fit_transform(list(air_t['Aircraft Manufacturer']))
Aircraft_Model = le.fit_transform(list(air_t['Aircraft Model']))
Landing_Count = le.fit_transform(list(air_t['Landing Count']))
Total_Landed_Weight = le.fit_transform(list(air_t['Total Landed Weight']))

In [26]:
# View ouput for one of the original categorical variables:
# ('Operating Airline IATA Code')
Op_IATA_Code

array([82, 82, 82, ..., 90, 90, 76], dtype=int64)

We can see that the output for 'Operating Airline IATA Code' is now numerical.

### Separate the data into features and labels

In [9]:
# labels: is a result we want
# features: the variables we can use to classify the 'Total Landed Weight'

In [11]:
# Define labels as (y) and features as (x)

# x list (features):      # (zip creates tuple objects)
X = list(zip(Operating_Airline, Op_IATA_Code, GEO_Summary, GEO_Region, Landing_Aircraft_Type, Aircraft_Body_Type, Aircraft_Manufacturer, Aircraft_Model, Landing_Count, Total_Landed_Weight))

# y list (labels):
y = list(Total_Landed_Weight)

### Split data into Train and Test sets

In [12]:
from sklearn.model_selection import train_test_split
# split the features and labels data (70/30)

# Test dataset:
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)  # test size ratio

In [13]:
# check the first 5 observations of the labels (y)
y[0:5]

[4823, 470, 3558, 384, 909]

### KNN Model

#### KNN Model 1

In [28]:
model = KNeighborsClassifier(n_neighbors=5)    # choose neighbours value

In [29]:
model.fit(x_train, y_train)          # train the model
acc = model.score(x_test, y_test)    # test accuracy of model
print("Model Accuracy is: ", acc)    # accuracy

Model Accuracy is:  0.3937815898299893


When neighbours is set to 5, model accuracy is 39.4%

#### KNN Model 2

In [17]:
model = KNeighborsClassifier(n_neighbors=7)    # choose neighbours value

In [18]:
model.fit(x_train, y_train)          # train the model
acc = model.score(x_test, y_test)    # test accuracy of model
print("Model Accuracy is: ", acc)    # accuracy

Model Accuracy is:  0.35794149180578955


When neighbours is set to 7, model accuracy is 35.8%

#### KNN Model 3

In [32]:
model = KNeighborsClassifier(n_neighbors=9)    # choose neighbours value

In [31]:
model.fit(x_train, y_train)          # train the model
acc = model.score(x_test, y_test)    # test accuracy of model
print("Model Accuracy is: ", acc)    # accuracy

Model Accuracy is:  0.2830448767039363


When neighbours is set to 9, model accuracy is 33.8%

From the results of the 3 models, as the neighbours value increases from 5 to 7 to 9, the model accuracy reduces.